runtime/compiler/z/codegen/J9S390CHelperLinkage.cpp

```diff
@@ -439,3 +459,537 @@ TR::Register * TR::S390CHelperLinkage::buildDirectDispatch(TR::Node * callNode,
+                                                  TR::Register* systemStackPointer,
+                                                  TR::Register* vmThreadRegister)
+   {
+   // CSP save offset is  JIT_GPR_SAVE_OFFSET(CSP) defined as:
+   // define({JIT_GPR_SAVE_OFFSET},{eval(STACK_BIAS+J9TR_cframe_jitGPRs+((CSP)*J9TR_pointerSize))})
+   //
+   // C stack pointer (CSP) is GPR15 on zLinux, and GPR4 on zOS
+   uint32_t cspNum = 4;
+   uint32_t stackBias  = 2048;
+   uint32_t cframeJitGPRs = offsetof(J9CInterpreterStackFrame, jitGPRs);
+
+   if (TR::Compiler->target.isLinux())
+      {
+      cspNum = 15;
+      stackBias = 0;
+      }
+
+   uint32_t cspSaveBias = stackBias + cframeJitGPRs + cspNum*sizeof(intptrj_t);
+
+   traceMsg(comp(), "Nigel CHelper linkage cspNum %d, stackBias %d, cframeJitGPRs %d, stack pointer overall bias %d\n",
+            cspNum, stackBias, cframeJitGPRs, cspSaveBias);
+
+   // store SSP to system stack SSP save slot
+   // CEEHDLR will hang if the SSP is not in the returnValue field of J9VMThread
+   generateSS1Instruction(cg(), TR::InstOpCode::MVC, callNode, sizeof(intptrj_t) - 1,
+                          generateS390MemoryReference(systemStackPointer, cspSaveBias, cg()),
+                          generateS390MemoryReference(vmThreadRegister, offsetof(J9VMThread, returnValue), cg()));
+
+   // The JNI definitly need GPR13 on the system stack for CEEHDLR.
+   // Otherwise, the GPR13 will be 0 after JNI call returns. This leads to crashes in the JIT'ed code.
+   uint32_t gpr13SaveOffset = stackBias + cframeJitGPRs + 13*sizeof(intptrj_t);
+   generateRXInstruction(cg(), TR::InstOpCode::getStoreOpCode(), callNode, vmThreadRegister,
+                         generateS390MemoryReference(systemStackPointer, gpr13SaveOffset, cg()));
+
+
+   static bool saveAllRegs = feGetEnv("jni_helper_save_all_regs") != NULL;
+
+   if (saveAllRegs)
+      {
+      uint32_t gprSaveBias = stackBias + cframeJitGPRs;
+      uint32_t hprSaveBias = stackBias + cframeJitGPRs + 16*sizeof(intptrj_t);
+
+      // Save R0 to R15 to c stack
+      generateRSInstruction(cg(), TR::InstOpCode::getStoreMultipleOpCode(), callNode,
+                            getS390RealRegister(REGNUM(1)), getS390RealRegister(REGNUM(16)),
+                            generateS390MemoryReference(systemStackPointer, gprSaveBias, cg()));
+
+      if (TR::Compiler->target.is32Bit() && cg()->supportsHighWordFacility())
+         generateRSInstruction(cg(), TR::InstOpCode::STMH, callNode,
+                               getS390RealRegister(TR::RealRegister::HPR0),
+                               getS390RealRegister(TR::RealRegister::HPR15),
+                               generateS390MemoryReference(systemStackPointer, hprSaveBias, cg()));
+
+      saveFPRsOrVRFs(callNode, systemStackPointer, vmThreadRegister);
+      }
+   }
+#endif
+
+TR::Register *
+TR::S390CHelperLinkage::buildIndirectDispatch(TR::Node * callNode,
```

Comment:
This function seems built very specifically for JNI IFA. Does it even belong here? There are also a handful of booleans below which can be folded away. I'm questioning the nature of the existence of this API. CHelper functions should all be statically addressable. Why exactly do we need indirect dispatch for them?

Reply:
#1 Admittedly this API is very customized towards the JNI IFA calls because IFA is the only consumer of this. Some of the booleans (`isFastPathOnly` and even the `#if defined(J9ZOS390)` for instance) can be folded away. I kept them there in case future users may need them. The IFA calls are zOS only but I figured that we can build indirect helper calls for other helpers on Linux.

#2 Also, the c function `fast_jitPost(pre)JNICallOffloadCheck` can't be addressed statically as per current design. Direct invocation will result in VM assert failures. I suspect all function pointers in the JITConfig structure are not directly invoke-able. By `directly invoke-able`, I mean BRASL jump to the `fast_jitPre(post)JNICallOffloadCheck` function pointers directly.

I didn't use `callNode->getSymbolReference()->getSymbol()->castToMethodSymbol()->getMethodAddress();` Will try and see how this is different from direct function pointer approach.

#3 One JNI IFA specific aspect is the register saving/restoring such as `saveRegistersForIFAHelper()`. This logic used to be baked inside the glue code is needed if CEEHDLR is enabled. CEE handlers can crash if sys SP and the vmthread registers are not saved on the native side before invoking the native. This has to be baked into the IFA call builder.

In conclusion:
#1 is simple to do. I can fold away the booleans.
For #2 we have to see why the C functions are not directly invoke-able using function pointers.
#3 has to be in the builder function, whether it be a CHelper function or a function in a different class.